### PR TITLE
Add callback

### PR DESCRIPTION
**Pull request by tmigot**
#101 Following the same model as https://github.com/JuliaSmoothOptimizers/JSOSolvers.jl/blob/main/src/R2.jl

### CODE REVIEW

File: src/method.jl [hunk 1]
```diff
@@ -280,30 +322,58 @@ function SolverCore.solve!(
...
+    done = stats.status != :unknown
   end
 
-  if solved
-    status = :first_order
+  set_solution!(stats, x)
+  stats
+end
+
+function get_status(
```
Comment (dpo) [inline src/method.jl:350]: Isn't this function being repeated in several packages? Should it go into SolverCore?

Comment (tmigot) [reply src/method.jl:350]: Yes, I think there are small differences though. But making a unified `get_status` in SolverCore would be great.
It is probably simpler if I make an issue in SolverCore to handle this separately.

File: src/method.jl [hunk 2]
```diff
@@ -90,6 +91,24 @@ The algorithm stops when ``‖c(xᵏ)‖ ≤ ctol`` and ``‖P∇L(xᵏ,λᵏ)
...
+The expected signature of the callback is `callback(nlp, solver, stats)`, and its output is ignored.
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
+All relevant information should be available in `nlp` and `solver`.
+Notably, you can access, and modify, the following:
+- `solver.x`: current iterate;
+- `solver.gx`: current gradient;
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.dual_feas`: norm of current projected gradient of Lagrangian;
+  - `stats.primal_feas`: norm of the feasibility;
```
Comment (dpo) [inline src/method.jl:105]:
```suggestion
  - `stats.primal_feas`: norm of the feasibility residual;
```

File: src/method.jl [hunk 3]
```diff
@@ -90,6 +91,24 @@ The algorithm stops when ``‖c(xᵏ)‖ ≤ ctol`` and ``‖P∇L(xᵏ,λᵏ)
...
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.multipliers`: current estimate of Lagrange multiplier associated to the equality constraint;
```
Comment (dpo) [inline src/method.jl:108]:
```suggestion
  - `stats.multipliers`: current estimate of Lagrange multiplier associated with the equality constraint;
```

**Review by dpo — APPROVED** [review]
Just two minor comments. Thanks!

**Pull request merged**